### ISSUE TITLE
picasa importer: 

### ISSUE THREAD

**Issue by bradfitz** [camlistore/camlistore#874]
I just got this error while importing my Google Photos...

> Nov 04 19:34:32 ante camlistored.run[29508]: 2016/11/04 19:34:32 picasa importer account, sha1-c08b33b46d579c37efe789586749973abd975d5e error: picasa importer: error importing album {1000000402159675 InstantUpload Auto Backup Auto Backup  https://picasaweb.google.com/115863474911002159675?alabel=small_instant_upload 2016-11-04 17:32:32 +0000 UTC 2016-11-04 18:16:25.372 +0000 UTC Brad Fitzpatrick https://picasaweb.google.com/xxx}: downloadAndParse(https://picasaweb.google.com/data/feed/api/user/default/albumid/xxx?imgmax=d&start-index=11001) got 500 Internal Server Error (Deprecated offset is too large for a stream ID query. Please  switch to using resume tokens.)

@mpl, can you take a look?

I just got my home server back online post-move, though, so my tree is a few months old, though. (I didn't want to update too many things at once)

**Comment by bradfitz** [comment]
I was at 95c4c9059

**Comment by mpl** [comment]
As @fgeller suggested in his e-mail, it looks like we're hitting:
https://code.google.com/p/gdata-issues/issues/detail?id=7004
and we're not the only ones:
https://github.com/google/gdata-java-client/issues/459

Given there's no feedback on the upstream issue, and there's no mention of that start-index limitation at https://developers.google.com/picasa-web/docs/2.0/reference I'm not sure where to go next, but I'll keep digging.

Also, https://developers.google.com/gdata/docs/2.0/reference#Queries mentions using a session-token, so maybe that's what the error message means by a resume token. I'll check how the other clients do it.

**Comment by mpl** [comment]
@bradfitz if you have any way of escalating or getting some feedback on https://code.google.com/p/gdata-issues/issues/detail?id=7004 , I think that could be useful.

**Comment by tgulacsi** [comment]
I see my photos in Google Drive. Maybe another importer from Google Drive could solve this?

**Comment by bradfitz** [comment]
@mpl, I've started to reach out to Google teams to see what the status is.

**Comment by bradfitz** [comment]
@tgulacsi, yup, that appears to be our only option at the moment. I created #896.

I filed a bug internally about the Picasaweb API but I all I heard was basically, "huh, that shouldn't happen", but I haven't see anybody racing to fix it either.

I think I'd prefer to work with the gdrive JSON discovery format API over the XML-GData API anyway.

**Comment by mholt** [comment]
@bradfitz @mpl This is probably nothing, but I did print out the response headers of a GET request and this header got my attention:

```
Set-Cookie: _rtok=zg-ry7QuH4Iw; Path=/; Secure; HttpOnly S=photos_html=0OkgeVSksOBDK8zFO8VX7IU0UQZnQTzb; Domain=.google.com; Path=/; Secure; HttpOnly
```

Does the `_rtok` cookie look like a "resume token"? I'll keep tinkering...

**Comment by bradfitz** [comment]
@mholt, interesting! Please report back either way.

**Comment by mholt** [comment]
@bradfitz Well, setting that cookie value in the next API request didn't work as a replacement for the start-index query string parameter, at least. _If_ it is a resume token, I don't know how to use it. 😛

**Comment by dfb** [comment]
Just a heads up on using Google Drive as a workaround: a number of people have a problem where not all of their photos from Google Photos show up in Google Drive. I know that someone mentioned wanting a Drive importer anyway, so I'm not trying to discourage that effort from moving ahead, but I did want to raise awareness that it may not provide a complete workaround for this Picasa issue.

I've been struggling with this exact issue for a month now and haven't come up with a good workaround, but out of 38k photos in Google Photos, approximately 2% are missing from Google Drive. Sometimes new  items in Photos take a few days to show up in Drive, other times they never show up at all.

I haven't found any way to contact the Picasa/Photos dev team and so I've just been going through the Drive support team, but it's been a month with no resolution and no solid info from them on what's wrong or if/when it'll be fixed. YMMV!

**Comment by mpl** [comment]
@dfb jsyk, we have a specific issue for doing google drive now: #896

**Comment by dfb** [comment]
Yup, thanks - I saw it but thought it's discussion would end up being more geared towards adding Drive support generally and not solely as a workaround to this problem, but please feel free to edit/copy the info as you see fit of course.

Anyway, I'm just passing through - I found this project and this ticket while trying to find a solution and figured I'd chime in so you guys know up front instead of making the sad and frustrating discovery after doing a bunch of work. ;-)

**Comment by mpl** [comment]
@dfb jsyk, we're hitting a similarly bad problem now: 

https://productforums.google.com/forum/#!msg/drive/HbNOd1o40CQ/VfIJCncyAAAJ

And while posting that, I've also seen all the people reporting what you reported: i.e. some of their photos not even getting synced at all to drive. a pretty terrible state of affairs indeed.

**Comment by dlipovetsky** [comment]
Following along here, the current state of things looks to be: I can make sure none of my albums have more than 10,000 images and use the Picasa importer, or I can make sure to not edit photos in Google Photos and use the Google Photos importer.

Apparently, Google Takeout  will give me all of my photos and metadata. I could download the Takeout data, then import. Even if the data is large, downloading to a GCP instance with sufficient local storage should work. I think the only piece missing is an Google Photos Takeout importer.

If you do not want to (or cannot) use the Picasa or Google Photos importers, would you consider using a Takeout importer? I think the biggest downside is that downloading and importing would not be incremental, but that may be acceptable if you can do both on a cloud instance.

**Comment by mpl** [comment]
I can see the interest of having a takeout importer in itself (not just for photos).

However, for photos specifically,
1) Do we even know if takeout will get you the latest version of your photos, and not, similarly to the bug in drive/photos, get you the initial version of your photo?
2) I would hope that Google someday fixes the drive API so that we can get the latest revision of the photo, which implicitly fixes the problem we have with the photos importer.

**Comment by dlipovetsky** [comment]
Thanks for the feedback. You make good points.

I just preformed an experiment that confirms that Takeout includes both original and edited versions of a photo (please see below for details).

I hope you are right that Google fixes the drive API. It seems safe to assume that, if and when the bug is fixed, the Google Photos importer will see the edited versions for the first time and import them. The risk is in not knowing how many photos I want to preserve in Camlistore are not originals, but edited versions. I'll start by using the Google Photos importer.

----
### Google Photos Takeout Experiment

1. Create album "Takeout Test"
2. Add photo to album
3. Edit photo (crop, color correct)
4. Add description to photo
5. Download album using Takeout

The results are [takeout-20171015T200254Z-001.tar.gz](https://github.com/camlistore/camlistore/files/1385788/takeout-20171015T200254Z-001.tar.gz).

I anonymized some values in the metadata, and removed an HTML page that was a personalized summary of the Takeout request. The page included the following description, which I found accurate:

> Your Google Photos data has been exported with the following formats:
> - Your photos and videos are exported in their original upload format.
> - Photos that have been edited are downloaded with the original file name ending with "-edited".
> - JSON - The photo metadata (data about the photo which may include a photo description or who is in >   the photo if available) is delivered in javascript object notation, which allows it to be easily parsed programmatically.

**Comment by mpl** [comment]
thanks for the experiment, that's great feedback.

What do you think about: we make a Google Takeout importer, and we add some sort of ignore/blacklist mechanism (like a .gitignore)? That way we do have a fully functional Google Takeout importer for those that want it, but you could also use it to import your photos only.

In any case, if the idea of a Google Takeout importer is interesting to you, please feel free to open a dedicated issue about, so we can track progress on that front specifically. (as I will close this issue here once the Google Photos importer has been tested enough and deemed working as expected).

**Comment by mpl** [comment]
Now that the importer supports deletions:

https://camlistore-review.googlesource.com/c/camlistore/+/11646 ,

one work-around to https://productforums.google.com/forum/#!msg/drive/HbNOd1o40CQ/VfIJCncyAAAJ

is to download your modified photo, remove it from Google Photos (or from Drive), and reupload the modified version to Google Photos. As it will be seen as a new photo (as it should), it will get synced to Drive, and the importer will pick it up (and it will remove the old/original version).

It sucks, but it's better than nothing.
